lib: json: Decoding of partially-recognized keys

BeckmaR opened the pull request:
IMO, the JSON library should be able to parse a JSON that contains known as well as unknown fields. These two tests verify that behaviour.

Fixes #55758

BeckmaR committed:
tests: lib: json: Decoding of partially-recognized keys

IMO, the JSON library should be able to parse a JSON that contains
known as well as unknown fields. These two tests verify that
behaviour.

Signed-off-by: René Beckmann <rene.beckmann@grandcentrix.net>

BeckmaR committed:
lib: os: json: Keep count of nesting depth

This fixes the bug where the parser just returns when encountering
an unexpected nested object.
Instead of always returning when encountering a `}` token, the
parser now keeps track of the "nesting depth".

Signed-off-by: René Beckmann <rene.beckmann@grandcentrix.net>

BeckmaR commented on tests/lib/json/src/main.c:
```diff
@@ -557,6 +557,28 @@ ZTEST(lib_json_test, test_json_key_not_in_descr)
+
+	ret = json_obj_parse(encoded, sizeof(encoded) - 1, test_descr,
+			     ARRAY_SIZE(test_descr), &ts);
+	zassert_equal(ret, 0b10, "some_int should be decoded");
```
Good idea, done

BeckmaR commented on lib/os/json.c:
```diff
@@ -627,9 +627,12 @@ static int64_t obj_parse(struct json_obj *obj, const struct json_obj_descr *desc
-		if (kv.value.type == JSON_TOK_OBJECT_END) {
+		if (kv.value.type == JSON_TOK_OBJECT_START) {
+			depth++;
+		} else if (kv.value.type == JSON_TOK_OBJECT_END && !--depth) {
```
Done

BeckmaR commented:
@martinjaeger Thanks for the review, I incorporated your suggestions and rebased.

martinjaeger reviewed (dismissed):
Thanks for the updates. LGTM.

MaureenHelm commented:
@M1cha can you take a look?

M1cha commented on lib/os/json.c:
```diff
@@ -627,10 +627,16 @@ static int64_t obj_parse(struct json_obj *obj, const struct json_obj_descr *desc
 	while (!obj_next(obj, &kv)) {
-		if (kv.value.type == JSON_TOK_OBJECT_END) {
-			return decoded_fields;
+		if (kv.value.type == JSON_TOK_OBJECT_START) {
+			depth++;
+		} else if (kv.value.type == JSON_TOK_OBJECT_END) {
+			depth--;
+			if (depth == 0) {
+				return decoded_fields;
+			}
 		}
```
mh doesn't this cause fields within the object of the unsupported field to be parsed with the description of the outer object?
In other words, what about:
```json
{
    "invalid": {"valid":6},
    "valid": 5
}
```

I didn't test this but looking at the code it seems like the resulting value of "valid" would be `6` instead of the expected `5`.
You might be able to get around this by `continuing` the loop when `depth > 1`.

BeckmaR replied:
You are right. I added a test that showcases this behavior. Very good eye as always.

BeckmaR replied:
Yeah, fun fact: This problem isn't new and already existed before my other fixes. I added another commit that fixes that problem, too.

mrfuchs commented:
Thanks for your PR and bringing this issue back up again.

A fix with a slightly different approach was addressed in PR #47989. The associated issue #47988 contains possible inputs to consider (cf. @M1cha's review above).

BeckmaR committed:
tests: lib: json: Test expected key in unexpected nested object

The nested object is unexpected and should be skipped completely,
but that is currently not the case. The value of some_int ends up
to be 42.

Signed-off-by: René Beckmann <rene.beckmann@grandcentrix.net>

BeckmaR committed:
lib: os: json: Ignore all fields in nested objects

This fixes a case where fields with expected names in unexpected
nested objects would be parsed incorrectly.

Example:
{
  "unexpected": {
    "expected": 42
  },
  "expected": 123
}

In this case, the parser should assign 123 to the top-level field
"expected", but would not, since it still read all the content
of the unexpected object.

Signed-off-by: René Beckmann <rene.beckmann@grandcentrix.net>

M1cha commented:
> Thanks for your PR and bringing this issue back up again.
> 
> A fix with a slightly different approach was addressed in PR #47989. The associated issue #47988 contains possible inputs to consider (cf. @M1cha's review above).

That approach might be better because if I'm not mistaken, the current approach would accept invalid json like:
```json
{
    "invalid": {{}},
    "valid": 5
}
```

At the same time, this PR would also not accept all valid inputs for the unsupported field, since we keep calling `obj_next` which parses the input as `"field": value` which will stop being true as soon as we enter other types like arrays:
```json
{
    "invalid": {
        "key": [0, 1, 2]
    },
    "valid": 5
}
```

I haven't tested any of these inputs (yet), but my brain compiler tells me these might be an issue.

mrfuchs commented:
> At the same time, this PR would also not accept all valid inputs for the unsupported field, since we keep calling `obj_next` which parses the input as `"field": value` which will stop being true as soon as we enter other types like arrays:
> 
> ```json
> {
>     "invalid": {
>         "key": [0, 1, 2]
>     },
>     "valid": 5
> }
> ```

Is this really the case? The `skip_field()` function in #47989 handles both `JSON_TOK_OBJECT_{START|END}` and `JSON_TOK_ARRAY_{START|END}` tokens and so should correctly handle the input you mention.

M1cha commented:
> > At the same time, this PR would also not accept all valid inputs for the unsupported field, since we keep calling `obj_next` which parses the input as `"field": value` which will stop being true as soon as we enter other types like arrays:
> > ```json
> > {
> >     "invalid": {
> >         "key": [0, 1, 2]
> >     },
> >     "valid": 5
> > }
> > ```
> 
> Is this really the case? The `skip_field()` function in #47989 handles both `JSON_TOK_OBJECT_{START|END}` and `JSON_TOK_ARRAY_{START|END}` tokens and so should correctly handle the input you mention.

with "this PR" I meant the PR you're currently commenting on. I've edited my post to make that more clear.

MaureenHelm commented:
@mrfuchs will you please take a look?
